Directly show "Forgot password?" link, fix #6808

jancborchardt opened the pull request:
We still need some JS so that when you click »Forgot password?«:

- [x] The password field is hidden
- [x] The log in button text is changed to »Reset password«
- [x] After the password is reset, the view is morphed back

Ref #6808 cc @ccoenen @nextcloud/javascript

jancborchardt committed:
Directly show Forgot password link, fix #6808

Signed-off-by: Jan-Christoph Borchardt <hey@jancborchardt.net>

ccoenen commented:
Why would this involve JavaScript at all? Ideally this would be a seperate url that I could point users to. As in "oh, I see, please go to *this URL* , enter your user name and you'll receive an email".

jancborchardt commented:
@ccoenen that is not mutually exclusive. :) The point is it’s a nicer experience for people from the log in page if it’s an animation rather than a hard page refresh. It can have a different URL though, like with a hash.

ccoenen commented on core/css/guest.css:
```diff
@@ -384,11 +384,21 @@ form .warning input[type='checkbox']+label {
 	margin: 10px 0;
 	text-align: center;
 	width: 100%;
+	text-shadow: 0 0 2px rgba(0, 0, 0, .4); // better readability on bright background
```
this only works accidentally, `//` is not a valid comment in regular CSS. (found once more below at line 572).

ccoenen commented:
when would this change in its current form have effect? I can see the `$_['canResetPassword']` in login.php, but as an outsider, I have no Idea what it actually means. I do not understand how/when the lost password container would be inserted.

jancborchardt committed:
Add some JS to hide and show elements for password reset

Signed-off-by: Jan-Christoph Borchardt <hey@jancborchardt.net>

jancborchardt commented:
Pushed some JS code to hide and show the elements for the password reset. Can someone with more technical knowledge check it please? @skjnldsv @blizzz @MorrisJobke @juliushaertl :)

ccoenen commented on core/templates/login.php:
```diff
@@ -51,7 +51,7 @@
 		</p>
 
-		<?php if (!empty($_['invalidpassword']) && !empty($_['canResetPassword'])) { ?>
+		<?php if (!empty($_['invalidpassword']) && !empty($_['canResetPlost-passwordassword'])) { ?>
```
this looks like a copy&paste error

jancborchardt replied:
Thanks, good catch.

jancborchardt committed:
Fix variable name

Signed-off-by: Jan-Christoph Borchardt <hey@jancborchardt.net>

jancborchardt commented:
This is ready to review now @nextcloud/javascript @rullzer @MorrisJobke @blizzz please check this out :)

skjnldsv reviewed (dismissed):
- When clicking the forgot password link then entering the username:
   - the submit button stays disabled.
   - using the enter key result by an error in the console: `An invalid form control with name='password' is not focusable.`
- When entering the username THEN clicking the forgot password link:
   - We could use a loader
   - The form is directly sent, I think we should probably wait for a manual action from the user

ccoenen commented:
I can't tell from the code alone (and I can't include this patch in my instance right now), can somebody tell me, if this contains a mechanism to _directly_ start which the lost password view? (Without the user needing to click or something)

Something, where I can send users to _directly_ reset their passwords. A URL i can stuff into an email easily. Something like these:

- https://www.paypal.com/authflow/password-recovery/
- https://de.wordpress.com/wp-login.php?action=lostpassword
- https://www.facebook.com/login/identify?ctx=recover&lwv=110
- https://fyp.ebay.de/EnterUserInfo

(you get the idea)

skjnldsv commented:
@ccoenen no, there is a few clicks needed.

ccoenen commented:
Thank you, in that case, I would like to request that this is added to this pull request as well. This is something I mentioned in the original ticket #6808, and that I think is rather important.

skjnldsv commented:
@ccoenen Well, it does directly show the "forgot password" button now, so it's just one click. You don't have to try to login to display the message now at least :)

jancborchardt commented:
As said, I would need a JS person to take it from here then. :)

skjnldsv committed:
Fix loading state and error message display

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv commented:
Okay, I'm still working on it, will finish when home from work:
- [ ] Message error css
- [ ] Loader fix and addition for lostpassword
- [ ] Toggle fix
- [ ] Fix the dual lostpassword buttons after a failed login
- [ ] The page reload on click after a failed login
- [ ] Fix disabled state of submit
- [ ] Fix enter key of form for lostpassword

skjnldsv committed:
Not disabled by default
Fixed transition bug of grouptop
Fix password submit and error handling
Fixed multiple form send
Fix submit loading state
fix login form sending state
Replace loader type

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv committed:
Fix template order and animation

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv commented:
Everything is fixed! :)
@nextcloud/designers @jancborchardt 

Honestly, we could really use a restructuration of the page, it looks a but messy. But it works and looks great ;)

Please check the following situations:
- Reset password with unknown user
- Reset with existing username
- Fail login then reset password with existing user
- Fail login then reset password with unknown user

Remember, every failed attempt will trigger a bruteforce protection, to test, remember to clean your `oc_bruteforce_attempts` table.

jancborchardt commented:
Thanks a lot @skjnldsv! :) Can we please have a review @nextcloud/javascript? :)

MorrisJobke reviewed (approved):
Tested and works 👍

MorrisJobke commented:
acceptance tests for login failed, but maybe it was a hiccup - I restated them (failed ones: https://drone.nextcloud.com/nextcloud/server/2642/239)

skjnldsv commented:
Nope, failure is related to this pr! :)

MorrisJobke reviewed (changes requested):
I get a "CSRF check failed" exception. Seems that you messed around with the hidden secret that is in the form to avoid such attacks. Please have a look

skjnldsv committed:
Fix acceptance test

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv commented:
Acceptance test passed, I restarted for an unrelated failure.

skjnldsv commented:
Few bugs on android
- [ ] too long button text
- [ ] Grey background when scrolled up

![screenshot_20171108-114305](https://user-images.githubusercontent.com/14975046/32544964-74c27896-c47a-11e7-90c0-08cb3403cc43.png)

MorrisJobke commented:
> I get a "CSRF check failed" exception. Seems that you messed around with the hidden secret that is in the form to avoid such attacks. Please have a look

@skjnldsv I have seen this after entering the new password

jancborchardt commented:
Should we merge for 13 and fix the issues separately? Cause they are mobile only.

MorrisJobke commented:
> Should we merge for 13 and fix the issues separately? Cause they are mobile only.

I'm fine with that.

jancborchardt commented:
@skjnldsv a fix for the first point you mentioned is at https://github.com/nextcloud/server/pull/7167. Second one I can’t reproduce, can you open an issue?

skjnldsv commented:
@jancborchardt Okay! :)

ccoenen commented:
thanks to everyone involved! Looking forward to the next release :-)
